### PR TITLE
Introduce RunContext object for passing necessary context to runner constructor methods

### PR DESCRIPTION
**Pull request by nkubala**
This change is a refactor that adds a new configuration object, `RunContext`, created by the runner. This object aggregates all contextual information about a skaffold pipeline run, and passes this object to all the constructors for the major components inside the runner (e.g. `Builder`, `Deployer`, etc.). This simplifies the constructor API, making it easier to add new parameters to these methods without having to change their signatures.

One small functional change: the `Tester` now optionally consumes the list of pre-built images (passed via the command line) from the `SkaffoldOptions` object, and uses this list to check against provided `TestCase`s from the skaffold.yaml when running tests. This is a cleaner way of skipping tests for pre-built images.

All other skaffold functionality should be unaffected by this change.

### CODE REVIEW

**Comment by balopat** [comment]
concerns: NeedsPush is not used anywhere, Plugins are failing on kokoro.

**Comment by balopat** [comment]
Plugins were failing due to [the gob encoder not supporting embedded structs with zero exported fields](https://github.com/golang/go/issues/5819). 
A solution to this is to refactor the `Build/Test/Deploy` parts to `Pipeline` in `SkaffoldConfig` and then only store `Pipeline` in `RunContext`:  [this experimental commit](https://github.com/GoogleContainerTools/skaffold/pull/1885/commits/cf8bd7a80d6f49caf402667287c8a2ac145e08b5) passes kokoro tests. I'll create a 3rd PR separately to tackle this change.

File: pkg/skaffold/test/types.go [hunk 1]
```diff
@@ -42,6 +42,7 @@ type Tester interface {
 // FullTester should always be the ONLY implementation of the Tester interface;
 // newly added testing implementations should implement the Runner interface.
 type FullTester struct {
+	blacklist  map[string]bool // a list of images that should not be tested
```
Comment (tejal29) [inline pkg/skaffold/test/types.go]: is this a relevant change? Can we drop this?

Comment (balopat) [reply pkg/skaffold/test/types.go]: yes

**Pull request merged**